### PR TITLE
4-Images Preview for Album

### PR DESCRIPTION
**Pull request by libasys**

### CODE REVIEW

**Comment by jancborchardt** [comment]
This is awesome! Much much better than the current solution. :+1: 

Please also review @icewind1991 @owncloud/designers

**Comment by jancborchardt** [comment]
@libasys are you coming to the ownCloud conference by the way? https://owncloud.org/conf you probably received an invitation already. :)

**Comment by jancborchardt** [comment]
@libasys just one thing: It doesn’t seem to use the full width. Sometimes 1 or 2 pixels are free to the right.

File: js/album.js [hunk 1]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
-		a.append(img);
-		img.height = targetHeight;
-		img.width = targetHeight * img.ratio;
+		var ratio = Math.round(img.ratio*100)/100;
+		var calcWidth = (targetHeight * ratio) / 2;
+		
+		a.width(calcWidth*2);
+		a.height(targetHeight-1);
+		
+		if(album.images.length > 1){
```
Comment (icewind1991) [inline js/album.js]: Ideally this would also take images in subalbums into account

**Comment by icewind1991** [comment]
The one problem I see with this is when an album as 2 or 3 pictures in it, not sure what the best way is to solve that case

**Comment by icewind1991** [comment]
For 2 we could maybe crop both previews to 100x200 px and fill the square with two images that way

**Comment by jbtbnl** [comment]
@icewind1991 and in case of 3 pictures, just pick two :)

**Comment by DeepDiver1975** [comment]
This approach looks good for 2 and 3 pics in one folder as well - let's merge this - we can still enhance from there on

@jancborchardt

File: js/album.js [hunk 2]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+			var Icount = album.images.length;
```
Comment (DeepDiver1975) [inline js/album.js]: camel case vars please

File: js/album.js [hunk 3]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+			if(Icount > 4) Icount=4;
```
Comment (DeepDiver1975) [inline js/album.js]: {} missing

File: js/album.js [hunk 4]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+			
+			for(var i=0; i<Icount; i++){
+				gm= new GalleryImage(album.images[i].src, 1);
```
Comment (DeepDiver1975) [inline js/album.js]: gm is undefined - var missing

File: js/album.js [hunk 5]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+				gm.getThumbnail(1).then(function (img) {
+					img[i] = img;
+					a.append(img[i]);
+				  	img[i].height = targetHeight / 2;
```
Comment (DeepDiver1975) [inline js/album.js]: whitespace mixxed

File: js/album.js [hunk 6]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+		a.width(calcWidth*2);
+		a.height(targetHeight-1);
+		
+		if(album.images.length > 1){
+			var Icount = album.images.length;
+			if(Icount > 4) Icount=4;
+			
+			for(var i=0; i<Icount; i++){
+				gm= new GalleryImage(album.images[i].src, 1);
+				gm.getThumbnail(1).then(function (img) {
```
Comment (DeepDiver1975) [inline js/album.js]: JSHInt: don't make functions within a loop

File: js/album.js [hunk 7]
```diff
@@ -23,16 +23,44 @@ Album.prototype.getThumbnailWidth = function () {
...
+				  	img[i].height = targetHeight / 2;
+				  	img[i].width = calcWidth;
+				  	aWidth = img[i].width;
+				});
+			}
+			
+		
+		}else{
+			a.append(img);
+		  	img.height = targetHeight;
```
Comment (DeepDiver1975) [inline js/album.js]: whitespace mixxed

**Comment by libasys** [comment]
wait i have added additional code for 4-images preview if a subalbum exist and on the album no images was found!

**Comment by requist** [comment]
:+1: Tested ok en looks most of the time good. I do find it becoming a bit messy when the pictures contain a lot of white. But probably drawing a very small border between the pictures will make it look too tightly organized.

**Comment by MorrisJobke** [comment]
> For 2 we could maybe crop both previews to 100x200 px and fill the square with two images that way

This could also be done for 3 pictures (the first on is 100x200 and the other two 100x100). We should do this in another PR.

**Comment by MorrisJobke** [comment]
I like this one and it works very well :+1:

**Comment by jancborchardt** [comment]
Thank you @libasys! :)

**Comment by jancborchardt** [comment]
@karlitschek I’d really like to backport this because it massively improves the way folders are displayed. Ok?

**Comment by karlitschek** [comment]
Yes please. Let's do an exception in this case.

**Comment by jancborchardt** [comment]
Thanks! Backported d971df7 f5eca5e fc974a8

**Pull request merged**